ddl/ddl.go

```diff
@@ -1103,40 +1107,81 @@ func (d *ddl) modifiable(origin *types.FieldType, to *types.FieldType) bool {
+func (d *ddl) getModifiableColumnJob(ctx context.Context, ident ast.Ident, spec *ast.AlterTableSpec) (*model.Job, error) {
 	is := d.infoHandle.Get()
 	schema, ok := is.SchemaByName(ident.Schema)
 	if !ok {
-		return errors.Trace(infoschema.ErrDatabaseNotExists)
+		return nil, errors.Trace(infoschema.ErrDatabaseNotExists)
 	}
 	t, err := is.TableByName(ident.Schema, ident.Name)
 	if err != nil {
-		return errors.Trace(infoschema.ErrTableNotExists)
+		return nil, errors.Trace(infoschema.ErrTableNotExists)
 	}
-	colName := spec.Column.Name.Name
+
+	colName := spec.NewColumn.Name.Name
 	col := table.FindCol(t.Cols(), colName.L)
 	if col == nil {
-		return infoschema.ErrColumnNotExists.GenByArgs(colName, ident.Name)
+		return nil, infoschema.ErrColumnNotExists.GenByArgs(colName, ident.Name)
 	}
-	if spec.Constraint != nil || spec.Position.Tp != ast.ColumnPositionNone ||
-		len(spec.Column.Options) != 0 || spec.Column.Tp == nil {
+	if spec.Constraint != nil || (spec.Position != nil && spec.Position.Tp != ast.ColumnPositionNone) ||
+		len(spec.NewColumn.Options) != 0 || spec.NewColumn.Tp == nil {
 		// Make sure the column definition is simple field type.
-		return errUnsupportedModifyColumn
+		return nil, errUnsupportedModifyColumn
 	}
-	d.setCharsetCollationFlenDecimal(spec.Column.Tp)
-	if !d.modifiable(&col.FieldType, spec.Column.Tp) {
-		return errUnsupportedModifyColumn
+	d.setCharsetCollationFlenDecimal(spec.NewColumn.Tp)
+	if !d.modifiable(&col.FieldType, spec.NewColumn.Tp) {
+		return nil, errUnsupportedModifyColumn
 	}
+
 	newCol := *col
-	newCol.FieldType = *spec.Column.Tp
+	newCol.FieldType = *spec.NewColumn.Tp
 	job := &model.Job{
 		SchemaID: schema.ID,
 		TableID:  t.Meta().ID,
 		Type:     model.ActionModifyColumn,
-		Args:     []interface{}{&newCol},
+		Args:     []interface{}{&newCol, col.Name},
+	}
+	return job, nil
+}
+
+// ChangeColumn renames an existing column and modifies the column's definition,
+// currently we only support limited kind of changes
+// that do not need to change or check data on the table.
+func (d *ddl) ChangeColumn(ctx context.Context, ident ast.Ident, spec *ast.AlterTableSpec) error {
+	if len(spec.NewColumn.Name.Schema.O) != 0 && ident.Schema.L != spec.NewColumn.Name.Schema.L {
+		return errWrongDBName.GenByArgs(spec.NewColumn.Name.Schema.O)
+	}
+	if len(spec.OldColumnName.Schema.O) != 0 && ident.Schema.L != spec.OldColumnName.Schema.L {
+		return errWrongDBName.GenByArgs(spec.OldColumnName.Schema.O)
 	}
+	if len(spec.NewColumn.Name.Table.O) != 0 && ident.Name.L != spec.NewColumn.Name.Table.L {
+		return errWrongTableName.GenByArgs(spec.NewColumn.Name.Table.O)
+	}
+	if len(spec.OldColumnName.Table.O) != 0 && ident.Name.L != spec.OldColumnName.Table.L {
+		return errWrongTableName.GenByArgs(spec.OldColumnName.Table.O)
+	}
+
+	newColName := spec.NewColumn.Name
+	spec.NewColumn.Name = spec.OldColumnName
+	job, err := d.getModifiableColumnJob(ctx, ident, spec)
```

Comment:
This is too tricky, add a `columnName` argument to `getModifiableColumnJob`, so we don't need to set `NewColumn.Name` to `OldColumnName` and later update job Args.